From 1f97e944090bd894f9bb00ce97afce94e86a62bd Mon Sep 17 00:00:00 2001
From: Timo Reichl <thetredev@gmail.com>
Date: Mon, 21 Apr 2025 01:06:11 +0200
Subject: [PATCH] core: declare some types as final

These types include `delete this` code, which C++17 doesn't like unless
the type is declared as `final`.

Signed-off-by: Timo Reichl <thetredev@gmail.com>
---
 core/metamod_plugins.cpp                    | 2 +-
 core/sourcehook/sourcehook.h                | 4 ++--
 core/sourcehook/sourcehook_impl_cvfnptr.cpp | 2 +-
 3 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/core/metamod_plugins.cpp b/core/metamod_plugins.cpp
index f33ff4a..92b4876 100644
--- a/core/metamod_plugins.cpp
+++ b/core/metamod_plugins.cpp
@@ -382,7 +382,7 @@ const char *CPluginManager::GetStatusText(CPlugin *pl)
 	}
 }
 
-struct Unloader : public SourceHook::Impl::UnloadListener
+struct Unloader final : public SourceHook::Impl::UnloadListener
 {
 	CPluginManager::CPlugin *plugin_;
 	bool destroy_;
diff --git a/core/sourcehook/sourcehook.h b/core/sourcehook/sourcehook.h
index ba01c12..5d9ae52 100644
--- a/core/sourcehook/sourcehook.h
+++ b/core/sourcehook/sourcehook.h
@@ -1039,7 +1039,7 @@ SourceHook::CallClass<T> *SH_GET_CALLCLASS(T *p)
 
 #define MAKE_DELEG(ret_type, params_decl, params_pass) \
 	struct IMyDelegate : ::SourceHook::ISHDelegate { virtual ret_type Call params_decl = 0; }; \
-	struct CMyDelegateImpl : IMyDelegate \
+	struct CMyDelegateImpl final : IMyDelegate \
 	{ \
 		FD m_Deleg; \
 		CMyDelegateImpl(FD deleg) : m_Deleg(deleg) {} \
@@ -1050,7 +1050,7 @@ SourceHook::CallClass<T> *SH_GET_CALLCLASS(T *p)
 
 #define MAKE_DELEG_void(params_decl, params_pass) \
 	struct IMyDelegate : ::SourceHook::ISHDelegate { virtual void  Call params_decl = 0; }; \
-	struct CMyDelegateImpl : IMyDelegate \
+	struct CMyDelegateImpl final : IMyDelegate \
 	{ \
 		FD m_Deleg; \
 		CMyDelegateImpl(FD deleg) : m_Deleg(deleg) {} \
diff --git a/core/sourcehook/sourcehook_impl_cvfnptr.cpp b/core/sourcehook/sourcehook_impl_cvfnptr.cpp
index 65ea49a..37b18ad 100644
--- a/core/sourcehook/sourcehook_impl_cvfnptr.cpp
+++ b/core/sourcehook/sourcehook_impl_cvfnptr.cpp
@@ -74,7 +74,7 @@ namespace SourceHook
 			return true;
 		}
 
-		class CVfnPtrOrigThunkCleanup : public ICleanupTask
+		class CVfnPtrOrigThunkCleanup final : public ICleanupTask
 		{
 			CPageAlloc *m_Allocator;
 			void *m_AddrToFree;
-- 
2.39.5

